File: apps/api/src/etag-test.ts
/**
 * ETag validation test for the TaskFlow API.
 *
 * This test starts the Express app and verifies that dynamic JSON
 * responses (e.g., /health) do NOT include an ETag header, confirming
 * that `app.disable("etag")` is working correctly.
 *
 * Usage: npx tsx src/etag-test.ts
 *        node --import tsx src/etag-test.ts
 */

import { createServer } from "http";
import express from "express";

// We cannot import the actual app because it calls listen() eagerly.
// Instead, we construct a minimal equivalent that replicates the exact
// setup from index.ts — including the `app.disable("etag")` call — and
// verify no ETag is emitted.
//
// This approach validates the principle: the same setting applied in
// index.ts must exist before route registration.

function buildApp(): express.Express {
  const app = express();
  // This mirrors the fix in index.ts — must appear BEFORE routes
  app.disable("etag");
  app.use(express.json());
  app.get("/health", (_req, res) => {
    res.json({ status: "ok", service: "taskflow-api" });
  });
  app.get("/users", (_req, res) => {
    res.json({ data: [], message: "stub" });
  });
  return app;
}

async function main(): Promise<void> {
  const app = buildApp();
  const server = createServer(app);

  // Bind to a random available port
  await new Promise<void>((resolve) => server.listen(0, resolve));
  const addr = server.address();
  if (!addr || typeof addr === "string") {
    console.error("FAIL: Could not determine server address");
    process.exit(1);
  }
  const port = (addr as any).port;
  console.log(`Test server listening on port ${port}`);

  let allPassed = true;

  for (const path of ["/health", "/users"]) {
    const res = await fetch(`http://127.0.0.1:${port}${path}`);
    const etag = res.headers.get("etag");
    const body = await res.json();

    if (etag) {
      console.error(`FAIL [${path}]: ETag header present: "${etag}"`);
      allPassed = false;
    } else {
      console.log(`PASS [${path}]: No ETag header`);
    }

    if (body && typeof body === "object") {
      console.log(`  Body keys: ${Object.keys(body).join(", ")}`);
    }
  }

  server.close();

  if (allPassed) {
    console.log("\n✓ All ETag validation tests passed.");
    process.exit(0);
  } else {
    console.error("\n✗ Some ETag validation tests failed.");
    process.exit(1);
  }
}

main();

File: apps/api/src/etag-test.py
#!/usr/bin/env python3
"""
ETag validation script for TaskFlow API.

Simplified version that directly validates the Express app configuration
by checking app settings, without starting a live server.

Usage: python3 apps/api/src/etag-test.py
"""

import re
import sys
import os


def test_index_ts_disables_etag():
    """Verify that apps/api/src/index.ts calls app.disable('etag') before routes."""
    index_path = os.path.join(
        os.path.dirname(__file__),
        "index.ts"
    )
    
    with open(index_path) as f:
        content = f.read()
    
    # Check that app.disable("etag") is present
    disable_str = 'app.disable("etag")'
    if disable_str not in content:
        print(f"FAIL: app.disable('etag') not found in {index_path}")
        return False
    print(f"PASS: app.disable('etag') found in index.ts")
    
    # Check that it appears BEFORE app.use(express.json()) (route registration)
    disable_pos = content.find('app.disable("etag")')
    if disable_pos == -1:
        disable_pos = content.find("app.disable('etag')")
    
    use_json_pos = content.find("app.use(express.json())")
    routes_start = content.find("app.get(")
    
    if disable_pos < use_json_pos and disable_pos < routes_start:
        print(f"PASS: app.disable('etag') is set before routes are registered")
    else:
        print(f"FAIL: app.disable('etag') must be set before route registration")
        return False
    
    return True


def test_health_endpoint_matches_expected():
    """Verify /health response body matches expected structure."""
    index_path = os.path.join(
        os.path.dirname(__file__),
        "index.ts"
    )
    
    with open(index_path) as f:
        content = f.read()
    
    # The /health handler must return { status: "ok", service: "taskflow-api" }
    if "status: \"ok\"" in content and "service: \"taskflow-api\"" in content:
        print("PASS: /health endpoint response body is preserved")
    else:
        print("FAIL: /health endpoint body has changed unexpectedly")
        return False
    
    return True


def main():
    tests = [
        ("ETag disabled", test_index_ts_disables_etag),
        ("Health endpoint preserved", test_health_endpoint_matches_expected),
    ]
    
    all_pass = True
    for name, fn in tests:
        print(f"\n--- {name} ---")
        if not fn():
            all_pass = False
    
    print()
    if all_pass:
        print("✓ All ETag validation tests passed.")
        sys.exit(0)
    else:
        print("✗ Some ETag validation tests failed.")
        sys.exit(1)


if __name__ == "__main__":
    main()
